### PR TITLE
Fix a sat mapper bug

### LINKED ISSUE #23: SATMapper fails on a specific instance
### Environment

- **Qiskit version**: 1.0.1
- **Python version**: 3.11.5
- **Operating system**: Windows

### What is happening?

The SATMapper fails on a random instance.

### How can we reproduce the issue?

```python
from qopt_best_practices.sat_mapping import SATMapper
from qiskit.transpiler.passes.routing.commuting_2q_gate_routing import SwapStrategy
import networkx as nx

#graph = nx.random_regular_graph(3, 6, seed=2)  # works
graph = nx.random_regular_graph(3, 6, seed=1)  # fails

swap_strategy = SwapStrategy.from_line([i for i in range(6)])

sat_mapper = SATMapper()
remapped_g, sat_map, min_sat_layers = sat_mapper.remap_graph_with_sat(
    graph=graph,
    swap_strategy=swap_strategy,
)
```
The trace is
```
File ...\qopt_best_practices\sat_mapping\sat_mapper.py:172, in SATMapper.remap_graph_with_sat(self, graph, swap_strategy)
    169 num_nodes = len(graph.nodes)
    170 results = self.find_initial_mappings(graph, swap_strategy, 0, num_nodes - 2)
--> 172 min_k = min((k for k, v in results.items() if v.satisfiable))
    173 edge_map = dict(results[min_k].mapping)
    175 remapped_graph = nx.relabel_nodes(graph, edge_map)

ValueError: min() arg is an empty sequence
```

### What should happen?

The SATMapper should not fail.

### Any suggestions?

None yet, need to dig deeper.

### PR DESCRIPTION
**Pull request by a-matsuo**
### Summary
Fix #23

### Details and comments
The SAT mapper did not compute the state of full connectivity, which should be (n-2). It only calculated up to (n-3).
Therefore, it raised an error for instances requiring full connectivity.

### CODE REVIEW

File: test/test_sat_mapping.py [hunk 1]
```diff
@@ -93,3 +93,14 @@ def test_deficient_strategy(self):
 
         # Crucially, if the `connectivity_matrix` in `find_initial_mappings` we get a wrong result.
         self.assertEqual(min_layer, 3)
+
+    def test_full_connectivity(self):
+        """Test that the SAT mapper works when the SWAP strategy has full connectivity."""
+        graph = nx.random_regular_graph(3, 6, seed=1)  # fails
```
Comment (ElePT) [inline test/test_sat_mapping.py:99]:
```suggestion
        graph = nx.random_regular_graph(3, 6, seed=1)
```

**Review by ElePT — COMMENTED** [review]
LGTM! I added a tiny comment

File: test/test_sat_mapping.py (hunk 1 above)
Comment (a-matsuo) [reply test/test_sat_mapping.py]: Thank you for pointing it out!

**Review by eggerdj — CHANGES_REQUESTED** [review]
Thanks Atsushi. Can we try and also cover the case of defective SWAP strategies, i.e., SWAP strategies that, no matter the number of layers, cannot reach full connectivity?

File: qopt_best_practices/sat_mapping/sat_mapper.py [hunk 2]
```diff
@@ -171,8 +171,7 @@ def remap_graph_with_sat(self, graph: nx.Graph, swap_strategy) -> tuple[int, dic
         graph gets mapped to node `v` in the Pauli strings.
         """
         num_nodes = len(graph.nodes)
-        results = self.find_initial_mappings(graph, swap_strategy, 0, num_nodes - 2)
-
+        results = self.find_initial_mappings(graph, swap_strategy, 0, num_nodes - 1)
         min_k = min((k for k, v in results.items() if v.satisfiable))
         edge_map = dict(results[min_k].mapping)
```
[the remark this answers is not in the document]
Comment (a-matsuo) [reply qopt_best_practices/sat_mapping/sat_mapper.py:176]: I understand. I have updated it to return None, None, None when the SAT mapper is unable to find solutions.

[the remark this answers is not in the document]
Comment (ElePT) [reply qopt_best_practices/sat_mapping/sat_mapper.py]: I think that returning `None, None, None` might be nicer to handle for users, as long as the behavior is clearly explained in the class docstring, so maybe that needs to be updated too?

[the remark this answers is not in the document]
Comment (a-matsuo) [reply qopt_best_practices/sat_mapping/sat_mapper.py]: Thank you for your comment! Yes, we definitely need to update the docstring. I will work on that.

File: test/test_sat_mapping.py [hunk 3]
```diff
@@ -93,3 +93,29 @@ def test_deficient_strategy(self):
...
+            swap_strategy=swap_strategy,
+        )
+        self.assertEqual(min_sat_layers, 4)
+
+    def test_unable_to_remap(self):
+        """Test that the SAT mapper works when the SWAP strategy is unable to remap."""
+        graph = nx.random_regular_graph(3, 6, seed=1)
+        swap_strategy = SwapStrategy.from_line(list(range(6)))
+        cmap = CouplingMap([(idx, idx + 1) for idx in range(5)])
+        swap_strategy = SwapStrategy(cmap, [])
```
Comment (eggerdj) [inline test/test_sat_mapping.py:113]:
```suggestion
        graph = nx.random_regular_graph(3, 6, seed=1)
        cmap = CouplingMap([(idx, idx + 1) for idx in range(5)])
        swap_strategy = SwapStrategy(cmap, [])
```
Looks like the first swap strat is redundent

**Review by eggerdj — COMMENTED** [review]
This looks good to me now. I have one last comment.

**Review by eggerdj — APPROVED** [review]
LGTM. Thanks a lot. Please fix whatever is failing in CI.

**Pull request merged**